From 779062b1dd6ef1f06fa6751ba1c53504d28e959b Mon Sep 17 00:00:00 2001
From: CHEN Gang <gang.c.chen@intel.com>
Date: Tue, 15 Oct 2013 11:35:40 +0800
Subject: [PATCH] Add more log for soft/hard lockup and watchdog reset

Now, when soft lockup happen, it will dump the call stack of current cpu core.

Found some cases like this:
One core disable interrupt for a long time.
This should be a hard lockup case. But at last, soft lockup happen.
The reason is one core run smp_call_function().
This function will disable preempt first, then generate IPI to other cores.
And this operation can be set as synchronized.
So the core disabled interrupt can not handle this IPI message.
Then soft lockup detect the issue before hard lockup.

For these kind of issue, the root cause is the core who disabled interrupt for a long time,
not the core run smp_call_function(). So only current core's call stack is not enough.

Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-15542
Change-Id: I794ad8dcff1797d04571c8879b323bf0f12e054e
Original Change-Id: Iad5ff70d47662c2aca5cd3c5720794a4360682f1
Signed-off-by: Pan Zhenjie <zhenjie.pan@intel.com>
Signed-off-by: CHEN Gang <gang.c.chen@intel.com>
Reviewed-on: http://android.intel.com:8080/138444
Reviewed-by: Benoit, NicolasX <nicolasx.benoit@intel.com>
Tested-by: Imougar, JacquesX <jacquesx.imougar@intel.com>
Reviewed-by: Imougar, JacquesX <jacquesx.imougar@intel.com>
Reviewed-by: Guerard, Christophe <christophe.guerard@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 kernel/watchdog.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/kernel/watchdog.c b/kernel/watchdog.c
index a404c05c0fea..02af46017a6b 100644
--- a/kernel/watchdog.c
+++ b/kernel/watchdog.c
@@ -314,8 +314,10 @@ static void watchdog_overflow_callback(struct perf_event *event,
 		if (__this_cpu_read(hard_watchdog_warn) == true)
 			return;
 
-		if (hardlockup_panic)
+		if (hardlockup_panic) {
+			trigger_all_cpu_backtrace();
 			panic("Watchdog detected hard LOCKUP on cpu %d", this_cpu);
+		}
 		else
 			WARN(1, "Watchdog detected hard LOCKUP on cpu %d", this_cpu);
 
@@ -401,8 +403,10 @@ static enum hrtimer_restart watchdog_timer_fn(struct hrtimer *hrtimer)
 		else
 			dump_stack();
 
-		if (softlockup_panic)
+		if (softlockup_panic) {
+			trigger_all_cpu_backtrace();
 			panic("softlockup: hung tasks");
+		}
 		__this_cpu_write(soft_watchdog_warn, true);
 	} else
 		__this_cpu_write(soft_watchdog_warn, false);
-- 
1.9.1

